browser/search_engines/search_engine_provider_service.cc

```diff
@@ -12,52 +15,64 @@
-#include "components/search_engines/template_url_service.h"
 
-SearchEngineProviderService::SearchEngineProviderService(
-    Profile* otr_profile)
-    : otr_profile_(otr_profile),
-      original_template_url_service_(
-          TemplateURLServiceFactory::GetForProfile(
-              otr_profile_->GetOriginalProfile())),
-      otr_template_url_service_(
-          TemplateURLServiceFactory::GetForProfile(otr_profile_)) {
+SearchEngineProviderService::SearchEngineProviderService(Profile* profile)
+    : profile_(profile),
+      template_url_service_(
+          TemplateURLServiceFactory::GetForProfile(profile_)) {
   use_alternative_search_engine_provider_.Init(
       kUseAlternativeSearchEngineProvider,
-      otr_profile_->GetOriginalProfile()->GetPrefs(),
+      profile_->GetOriginalProfile()->GetPrefs(),
       base::Bind(&SearchEngineProviderService::OnPreferenceChanged,
                  base::Unretained(this)));
 
-  std::vector<TemplateURLPrepopulateData::BravePrepopulatedEngineID>
-      alt_search_providers = {
-          TemplateURLPrepopulateData::PREPOPULATED_ENGINE_ID_DUCKDUCKGO,
-          TemplateURLPrepopulateData::PREPOPULATED_ENGINE_ID_DUCKDUCKGO_DE,
-          TemplateURLPrepopulateData::
-              PREPOPULATED_ENGINE_ID_DUCKDUCKGO_AU_NZ_IE
-      };
+  std::unique_ptr<TemplateURLData> data =
+      brave::GetDDGTemplateURLData(profile->GetPrefs());
+  alternative_search_engine_url_.reset(new TemplateURL(*data));
+  observation_.Observe(template_url_service_);
+}
+
+SearchEngineProviderService::~SearchEngineProviderService() = default;
+
+void SearchEngineProviderService::OnTemplateURLServiceChanged() {
+  // When this change is came from pref changing, we don't need to control
+  // prefs again.
+  if (ignore_template_url_service_changing_)
+    return;
 
-  std::unique_ptr<TemplateURLData> data;
-  for (const auto& id : alt_search_providers) {
-    data = TemplateURLPrepopulateData::GetPrepopulatedEngine(
-        otr_profile->GetPrefs(), id);
-    if (data)
+  // The purpose of below code is toggling alternative prefs
+  // when user changes from ddg to different search engine provider
+  // (or vice versa) from settings ui.
+  bool is_ddg_is_set = false;
+  switch (template_url_service_->GetDefaultSearchProvider()
+              ->data()
+              .prepopulate_id) {
+    case TemplateURLPrepopulateData::PREPOPULATED_ENGINE_ID_DUCKDUCKGO:
+    case TemplateURLPrepopulateData::PREPOPULATED_ENGINE_ID_DUCKDUCKGO_DE:
+    case TemplateURLPrepopulateData::PREPOPULATED_ENGINE_ID_DUCKDUCKGO_AU_NZ_IE:
+      is_ddg_is_set = true;
+      break;
+    default:
       break;
   }
 
-  // There should ALWAYS be one entry
-  DCHECK(data);
-  alternative_search_engine_url_.reset(new TemplateURL(*data));
-}
-
-SearchEngineProviderService::~SearchEngineProviderService() {
+  if (UseAlternativeSearchEngineProvider() || is_ddg_is_set)
+    brave::ToggleUseAlternativeSearchEngineProvider(profile_);
```

Comment:
Does this change `kUseAlternativeSearchEngineProvider` from true to false when user change the search engine setting from DDG to something else in settings?
<img width="706" alt="Screen Shot 2021-04-16 at 14 09 23" src="https://user-images.githubusercontent.com/11330831/115084222-6a0a4e80-9ebd-11eb-8f00-89a7b176b36e.png">

Reply:
Yes